sapicore/pipeline/simulation.py

```diff
@@ -0,0 +1,172 @@
+""" Simulate a network from a configuration YAML. """
+import os
+import logging
+
+from argparse import ArgumentParser
+from datetime import datetime
+
+import torch
+from torch import Tensor
+from alive_progress import alive_bar
+
+from sapicore.pipeline import Pipeline
+from sapicore.model import Model
+from sapicore.engine.network import Network
+
+from sapicore.utils.seed import fix_random_seed
+from sapicore.utils.signals import extend_input_current
+from sapicore.utils.tensorboard import TensorboardWriter, HDFData
+from sapicore.utils.constants import DT, TIME_FORMAT, SEED
+from sapicore.utils.io import ensure_dir, save_yaml, log_settings
+
+from sapicore.tests import ROOT
+
+# default hardware device for this pipeline.
+DEVICE = "cpu" if not torch.cuda.is_available() else "cuda:0"
```

Comment:
Should this be in constants? But I'm worried about setting the device on import. Because the device may not be ready when imported or whatever. So it may be better if we decide on the device later on.

Even better, is if you default a class variable to this value. Which would allow someone to overwrite the device before calling run. Or make it part of the config. Right now there isn't a way to set the device manually.

Reply:
All good ideas (could also add an option to override with a command line argument). Your class variable suggestion wins out in my opinion.

Reply:
Fixed:

```python
super().__init__(configuration=configuration, **kwargs)

if self.configuration and self.configuration.get("device"):
    # set hardware device from configuration.
    self.device = self.configuration.get("device")

else:
    # set default hardware device for this pipeline.
    self.device = "cpu" if not torch.cuda.is_available() else "cuda:0"
```